Added support for DBMS-based artifact store

avflor opened the pull request:
This PR adds support for using an RDBMS as an artifact store. Artifacts are stored in binary format. SQL Alchemy is used to connect to the RDBMS.

Commit:
1st version of DB artifact store

Commit:
Added list_artifact method and tests

Commit:
Added download artifacts methods

Commit:
Changes to support root path after db_uri

Commit:
Small changes in artifact repo registry and utils

Commit:
Fixed tests

Commit:
Merge branch 'master' into dbArtifactStore

Commit:
Fixed enum equality issues

apurva-koti commented on .travis.yml:
```diff
@@ -23,68 +23,68 @@ matrix:
-      install:
-        - INSTALL_LARGE_PYTHON_DEPS=true source ./travis/install-common-deps.sh
-      script:
-        - ./travis/run-large-python-tests.sh
```
What exactly is the reason for this change?

avflor replied:
@apurva-koti It is temporary just to speed up testing in travis. I will revert it once done.

Commit:
Removed comments from travis.yml

mateiz commented on mlflow/store/dbmodels/initial_artifact_store_models.py:
```diff
@@ -0,0 +1,66 @@
+Base = declarative_base()
+
+
+class SqlArtifact(Base):
```
Will this be added to all MLflow SQL store databases? Or do users have to opt in?

avflor replied:
@mateiz Do you mean the backend SQL database where runs are recorded? This table will not be added there. Only when the users specify a RDBMS as an artifact store, this table will be created in the RDBMS.

mateiz commented:
Thanks for starting this, Avrilia! Can you also add some docs about it so that people can see how to use it? You can edit the docs in the docs folder and include the changes in the same pull request (see the README there for how to build them).

avflor commented:
Thanks @mateiz! Sure, I will add the documentation once I fix the failing tests on travis.

Commit:
Added first draft of documentation

avflor commented:
@smurching @dbczumar Please let me know if you need anything from my side before having someone review this PR. Looking forward to your comments.

dbczumar commented on mlflow/store/db_artifact_repo.py:
```diff
@@ -0,0 +1,279 @@
+def extract_db_uri_and_path(artifact_uri):
+    parsed_uri = urllib.parse.urlparse(artifact_uri)
+
+    # for testing
```
Let's remove this code

dbczumar commented on mlflow/tracking/artifact_utils.py:
```diff
@@ -53,14 +58,40 @@ def _download_artifact_from_uri(artifact_uri, output_path=None):
+            artifact_path=artifact_path, dst_path=output_path)
+
+
+def extract_db_uri_and_artifact_path_from_uri(artifact_uri):
```
Do we need a new db uri extraction method here, or can we move the logic into `extract_db_uri_and_path`? Why do we parse additional information about the URI scheme here but not in `extract_db_uri_and_path`?

dbczumar commented on mlflow/store/db_artifact_repo.py:
```diff
@@ -0,0 +1,279 @@
+    return relative_path_to_artifact_path(rel_path) if rel_path is not None else None
+
+
+def extract_db_uri_and_path(artifact_uri):
```
Can you add unit tests for this method for multiple DB URIs with multiple schemes (`mssql` and `sqlite`), varying artifact path lengths, and additional URI components (dialects, drivers, params)? 

The existing unit tests only seem to address vanilla `sqlite://` URIs .

dbczumar commented on mlflow/store/db_artifact_repo.py:
```diff
@@ -0,0 +1,279 @@
+            return local_path
+
+        if artifact_path:
+            artifact_path = os.path.normpath(artifact_path)
```
Is there a particular reason for using `normpath` here rather than `abspath`? Otherwise, the entire `download_artifacts` method seems identical to `ArtifactRepository.download_artifacts` and can probably be removed, provided `normpath` is not necessary.

dbczumar replied:
@avflor Can you provide input here?

avflor replied:
The `ArtifactRepository.download_artifacts` does not apply `normpath `or `abspath `to the input `artifact_path` but only to `dst_path `. I've tried using the default one but I get test failures on windows. Example error:

```
> self = <tests.store.test_db_artifact_repo.TestSqlAlchemyStoreSqlite testMethod=test_download_file_artifact>
> 
>     def test_download_file_artifact(self):
> 
>         with TempDir() as root_dir:
> 
>             with open(root_dir.path("file_one.txt"), "w") as f:
> 
>                 f.write('DB store Test One')
> 
>     
> 
>             self.store.log_artifacts(root_dir._path, 'new_path/path')
> 
>             local_path = self.store.download_artifacts(artifact_path='new_path/path/file_one.txt')
> 
> >           assert open(local_path).read() == 'DB store Test One'
> 
> E           FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\travis\\AppData\\Local\\Temp\\tmpb8oe5tyk\\file_one.txt'
> 
> tests\store\test_db_artifact_repo.py:186: FileNotFoundError
```

> 
> 
> @avflor Can you provide input here?

dbczumar reviewed (changes requested):
@avflor I tried running the following sample code and ran into issues:

```
import mlflow
from mlflow.tracking.client import MlflowClient

client = MlflowClient()
exp_name = "artifact_test_experiment"
client.create_experiment(exp_name, artifact_location="sqlite:///testartifactdb")
mlflow.set_experiment(exp_name)

fpath = "test.txt"
with open(fpath, "w") as f:
    f.write("TEST")

with mlflow.start_run():
    mlflow.log_artifact(fpath, "test_file")
```

Initially, before commenting out the `# for testing` code in `db_artifact_repo`, this code failed with the following stacktrace:
```
Traceback (most recent call last):
  File "logtest.py", line 14, in <module>
    mlflow.log_artifact(fpath, "test_file")
  File "/Users/czumar/mlflow/mlflow/tracking/fluent.py", line 266, in log_artifact
    MlflowClient().log_artifact(run_id, local_path, artifact_path)
  File "/Users/czumar/mlflow/mlflow/tracking/client.py", line 243, in log_artifact
    artifact_repo = get_artifact_repository(run.info.artifact_uri)
  File "/Users/czumar/mlflow/mlflow/store/artifact_repository_registry.py", line 115, in get_artifact_repository
    return _artifact_repository_registry.get_artifact_repository(artifact_uri)
  File "/Users/czumar/mlflow/mlflow/store/artifact_repository_registry.py", line 76, in get_artifact_repository
    return repository(artifact_uri)
  File "/Users/czumar/mlflow/mlflow/store/db_artifact_repo.py", line 77, in __init__
    insp = sqlalchemy.inspect(self.engine)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/inspection.py", line 63, in inspect
    ret = reg(subject)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/reflection.py", line 140, in _insp
    return Inspector.from_engine(bind)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/reflection.py", line 136, in from_engine
    return Inspector(bind)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/reflection.py", line 110, in __init__
    bind.connect().close()
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 2193, in connect
    return self._connection_cls(self, **kwargs)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 103, in __init__
    else engine.raw_connection()
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 2293, in raw_connection
    self.pool.unique_connection, _connection
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 2266, in _wrap_pool_connect
    e, dialect, self
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1536, in _handle_dbapi_exception_noconnection
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 383, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 2262, in _wrap_pool_connect
    return fn()
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/pool/base.py", line 294, in unique_connection
    return _ConnectionFairy._checkout(self)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/pool/base.py", line 751, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/pool/base.py", line 483, in checkout
    rec = pool._do_get()
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/pool/impl.py", line 237, in _do_get
    return self._create_connection()
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/pool/base.py", line 299, in _create_connection
    return _ConnectionRecord(self)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/pool/base.py", line 428, in __init__
    self.__connect(first_connect_check=True)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/pool/base.py", line 630, in __connect
    connection = pool._invoke_creator(self)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/strategies.py", line 114, in connect
    return dialect.connect(*cargs, **cparams)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 453, in connect
    return self.dbapi.connect(*cargs, **cparams)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) unable to open database file
(Background on this error at: http://sqlalche.me/e/e3q8)
```

This appears to be caused by initializing a nonexistent database with URI `sqlite:///testartifactdb/cd0634a624c04f848cb26c8218b2d218/artifacts`.

After removing the `# for testing` code, I'm still running into errors parsing the database URI:
```
Traceback (most recent call last):
  File "logtest.py", line 14, in <module>
    mlflow.log_artifact(fpath, "test_file")
  File "/Users/czumar/mlflow/mlflow/tracking/fluent.py", line 266, in log_artifact
    MlflowClient().log_artifact(run_id, local_path, artifact_path)
  File "/Users/czumar/mlflow/mlflow/tracking/client.py", line 243, in log_artifact
    artifact_repo = get_artifact_repository(run.info.artifact_uri)
  File "/Users/czumar/mlflow/mlflow/store/artifact_repository_registry.py", line 115, in get_artifact_repository
    return _artifact_repository_registry.get_artifact_repository(artifact_uri)
  File "/Users/czumar/mlflow/mlflow/store/artifact_repository_registry.py", line 76, in get_artifact_repository
    return repository(artifact_uri)
  File "/Users/czumar/mlflow/mlflow/store/db_artifact_repo.py", line 74, in __init__
    self.engine = sqlalchemy.create_engine(self.db_uri)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/__init__.py", line 455, in create_engine
    return strategy.create(*args, **kwargs)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/strategies.py", line 54, in create
    u = url.make_url(name_or_url)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/url.py", line 229, in make_url
    return _parse_rfc1738_args(name_or_url)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/url.py", line 291, in _parse_rfc1738_args
    "Could not parse rfc1738 URL from string '%s'" % name
sqlalchemy.exc.ArgumentError: Could not parse rfc1738 URL from string 'sqlite:/testartifactdb/cd0634a624c04f848cb26c8218b2d218/artifacts'
```

Can you take a look at this and introduce some test coverage for MLflow's client/fluent `log_artifacts` APIs against the new DBMS-based artifact store?

avflor commented:
Hi @dbczumar . Thanks for the comments. There is a problem supporting sqllite in the dbartifactstore and I believe this is the reason why your code is failing. I ran your code with sqlserver and it works fine.

The db_artifact_store_repo takes as input a DB_URI of the form: <dialect>+<driver>://<username>:<password>@<host>:<port>/<database>?<params> and parses it to identify how to connect to the db. Sqllite has a weird syntax allowing for sqllite:///file_path expressions. In your example, to connect to the db I would need just the 'sqlite:/testartifactdb' path but what comes eventually to the db_artifact_store is a URI of the form 'sqlite:/testartifactdb/cd0634a624c04f848cb26c8218b2d218/artifacts' as you can see in the error message produced above. In this case, I do not know which path corresponds to db and which corresponds to the file path under which I should store the artifact. This is because mlflow extends the path before sending it to the DB-artifact store. Note that I don't have this problem with other DBs as they contain more info in the URI path such as drivers etc that help me navigate correctly. For example:

1) URI parsed by urlparse when using SQL Server: ParseResult(scheme='mssql+pyodbc', netloc='sqluser:xxx@xxx@yyy:000', path='/mlflow_test', params='',query='driver=ODBC+Driver+17+for+SQL+Server/9c7948ec0a714dbbb6acbd3096650665/artifacts', fragment='')

In this one the combination of path and query help understand what is db_path vs root_directory for artifacts.

2) URI parsed by urlparse when using sqllite: ParseResult(scheme='sqlite', netloc='', path='/testartifactdb/cb31d7d2ca7b4744b270d24bbd42da31/artifacts', params='', query='', fragment='')

In this one, I don't have the necessary info to differentiate.

Thus, I used sqlite only for testing the module at the artifact_store level. (Note that the tests still pass when I used SQL server) . Because of this problem we can go two ways: 1) Do not support sqllite or 2) Introducing some assumptions at the artifact store level (e.g., experimentid/artifacts is the root directory under which the artifacts are always stored) . Let me know what you think.

Commit:
Incorporated root_uri in the dbartifactStoreRepo

Commit:
Added support for sqlite URIs

avflor commented:
@dbczumar I fixed the issues with sqllite so your test works on my environment now for both sqlite and sqlserver. After the additions, some of your comments above do not apply anymore. I also added tests for the extract_uri method to cover multiple types of URIs. Finally, I added comments in the code to clarify the changes. Please let me know if you have more questions.

avflor commented:
@dbczumar Any comments on the PR after the changes? Please let me know if anything else is needed. Thanks!

avflor commented:
@dbczumar It's been a while since I addressed your previous comments. Any update on this PR?  People at Microsoft are excited about using a database as artifact store and we would really love to have this feature being part of an upcoming mlflow release.

dbczumar commented:
@avflor So sorry for the delay! I'll give this a thorough review within the next 24 hours. Thank you for your contribution! This is definitely an exciting feature.

dbczumar commented on mlflow/store/db_artifact_repo.py:
```diff
@@ -0,0 +1,304 @@
+                    session.query(SqlArtifact).filter(
+                        or_(SqlArtifact.group_path.like(
+                            os.path.normpath(os.path.join(self.root, regex))),
+                            SqlArtifact.group_path == os.path.normpath(
```
Is it possible to refactor this so that only the `like` clause is needed? Combined with the change suggested above, this might look something like:

```
        path = os.path.join(self.root, path) if path is not None else self.root
        with self.ManagedSessionMaker() as session:
            return [artifact.to_file_info()
                    for artifact in
                    session.query(SqlArtifact).filter(
                        SqlArtifact.group_path.like(
                            path + "%"))]
```

avflor replied:
This does not work in various edge cases. For example, one might ask for artifacts under /artifacts/path and get also artifact under /artifacts/path2

dbczumar commented on mlflow/store/dbmodels/initial_artifact_store_models.py:
```diff
@@ -0,0 +1,66 @@
+        :return: :py:class:`mlflow.entities.FileInfo`.
+        """
+        return FileInfo(
+            path=os.path.join(self.group_path, self.artifact_name),
```
```suggestion
            path=self.artifact_name,
```

This will allow the `FileInfo` results to conform to the expected semantics of the `ListArtifacts` API. MLflow expects `list_artifacts` to return file paths *relative* to the specified input path.

avflor replied:
@dbzcumar, I'm a bit confused with this comment, Assume the full path is
db_uri/runid/artifacts/subir/filename

What part of it should be captured in the fileinfo object? With the change you are suggesting fileinfo will just contain the filename part (not the subdir). Is that the expected behavior?

dbczumar replied:
@avflor Correct, `path` should only contain the filename part.

avflor replied:
@dbczumar Thanks for your prompt reply, but I'm still not sure this is the correct behavior. As you said the `FileInfo `is returned by `list_artifacts` which is in turn invoked by `download_artifacts`. If you take a look at the `download_artifacts` in artifact_repo.py  it expects as input a relative source path to the desired artifact ( I assume that this means relative to the root). This function invokes the `download_artifacts_into` function which also takes as input a relative path. The latter function is recursive and every time it is invoked the relative path is instantiated by the FileInfo object. It seems that by passing just the filename in this object this download will not work. Am I missing sth here?

dbczumar replied:
@avflor Sorry for the confusion. You are correct. The current behavior of `to_file_info()` is correct. Unfortunately, the UI is unable to display artifacts. I'll spend some time trying to debug that issue independently of `to_file_info()`.

avflor replied:
@dbczumar Sounds good. Is the ui problem only related to this PR or do you see issues with other types of artifact repos? I'm thinking whether the problem might be on how we define "relative_path" in the context of db_artifact_repo. For example if the full path is:

 db_uri/runid/artifacts/subdir/filename

my Fileindo currently returns "run_id/artifacts/subdir/filename". I wonder whether it should return "subdir/filename".

dbczumar replied:
@avflor The UI problem only seems to occur when using `DBArtifactRepository` from this PR. I think you've correctly identified the issue. `FileInfo` should return `subdir/filename`. One possible solution is to add a `base_path` argument to `to_file_info()` as follows:
```
    def to_file_info(self, base_path):
        """
        Convert DB model to corresponding FileInfo object.

        :return: :py:class:`mlflow.entities.FileInfo`.
        """
        return FileInfo(
            path=os.path.relpath(path=os.path.join(self.group_path, self.artifact_name), start=base_path),
            is_dir=False,
            file_size=self.artifact_initial_size)
```
You can then invoke `to_file_info(path)` in `list_artifacts`:
```
        path = os.path.join(self.root, path) if path is not None else self.root
        with self.ManagedSessionMaker() as session:
            return [artifact.to_file_info(path)
                    for artifact in
                    session.query(SqlArtifact).filter(
                        SqlArtifact.group_path.like(
                            path + "%"))]
```

Thanks for catching this!

avflor replied:
It seems that with this change the support of None in the `list_artifacts`, the UI works now.

dbczumar reviewed (commented):
@avflor I left a few comments that should fix some of the UI issues you were running into when preparing for your demo. I'll send information about how to structure this code as a plugin, rather than a commit to the main MLflow repo, this evening. This will provide your organization with more control over the code and functionality as you continue to expand the feature set.

Commit:
fixed conflicts with master

Commit:
addressed review comments

avflor commented on mlflow/store/db_artifact_repo.py:
```diff
@@ -0,0 +1,304 @@
+    parsed_uri = urllib.parse.urlparse(repo_uri)
+    scheme = parsed_uri.scheme
+    scheme_plus_count = scheme.count('+')
+    if scheme_plus_count != 0 and scheme_plus_count != 1:
```
Done

avflor commented on mlflow/store/db_artifact_repo.py:
```diff
@@ -0,0 +1,304 @@
+
+        :return: List of artifacts as FileInfo listed directly under path.
+        """
+        path = os.path.normpath(path)
```
The function now supports None as path.

aarondav commented:
A DBMS-based artifact store is now available as an external plugin: https://mlflow.org/docs/latest/plugins.html#sql-server-plugin
